deepchem/models/torch_models/modular.py

```diff
@@ -342,3 +271,93 @@ def fit_generator(self,
+        Specific components can be chosen by passing a list of strings with the desired component names. If both a source_model and a checkpoint/model_dir are loaded, the source_model weights will be loaded.
+
+        Parameters
+        ----------
+        source_model: dc.ModularTorchModel, required
+            source_model can either be the pretrained model or a dc.TorchModel with
+            the same architecture as the pretrained model. It is used to restore from
+            a checkpoint, if value_map is None and to create a default assignment map
+            if assignment_map is None
+        checkpoint: str, default None
+            the path to the checkpoint file to load.  If this is None, the most recent
+            checkpoint will be chosen automatically.  Call get_checkpoints() to get a
+            list of all available checkpoints
+        model_dir: str, default None
+            Restore source model from custom model directory if needed
+        inputs: List, input tensors for model
+            if not None, then the weights are built for both the source and self.
+        """
+        if inputs is not None:
+            # Ensure weights for both models are built.
+            if source_model:
+                source_model.model(inputs)
+            self.model(inputs)
+
+        self._ensure_built()
+
+        if source_model is not None:
+            for name, module in source_model.components.items():
+                if components is None or name in components:
+                    self.components[name] = module
+            self.build_model()
+
+        elif source_model is None:
+            self.restore(components=components,
+                         checkpoint=checkpoint,
+                         model_dir=model_dir)
+
+    def save_checkpoint(self, max_checkpoints_to_keep=5, model_dir=None):
+        if model_dir is None:
+            model_dir = self.model_dir
+        if not os.path.exists(model_dir):
+            os.makedirs(model_dir)
+
+        data = {}
+        data['model'] = self.model.state_dict()
+        for name, component in self.components.items():
+            data[name] = component.state_dict()
+
+        temp_file = os.path.join(model_dir, 'temp_checkpoint.pt')
+        torch.save(data, temp_file)
+
+        # Rename and delete older files.
+
+        paths = [
+            os.path.join(model_dir, 'checkpoint%d.pt' % (i + 1))
+            for i in range(max_checkpoints_to_keep)
+        ]
+        if os.path.exists(paths[-1]):
+            os.remove(paths[-1])
+        for i in reversed(range(max_checkpoints_to_keep - 1)):
+            if os.path.exists(paths[i]):
+                os.rename(paths[i], paths[i + 1])
+        os.rename(temp_file, paths[0])
+
+    def restore(  # type: ignore
+            self,
+            components: Optional[List[str]] = None,
+            checkpoint: Optional[str] = None,
+            model_dir: Optional[str] = None) -> None:
```

Comment:
Let's add a docstring here that explains how restore handles component checkpointing. The raw code is confusing